source/slang/slang-check-decl.cpp

```diff
@@ -9481,8 +9484,9 @@ namespace Slang
             ensureDecl(visitor, referencedDecl, DeclCheckState::CapabilityChecked);
         }
         
-        if (resultCaps.implies(nodeCaps))
-            return;
+        // Always join capabilities to ensure only the lowest denominator will be apart of the final requirments.
```

Comment:
There may be a subtlety here, but the intention here is to reduce the amount of calls to `join` which can be expensive.
Generally, we will if resultCaps = (compute|fragment) and nodeCaps = (fragment), then resultCaps shouldn't imply nodeCaps, so this early return should never hit?

Reply:
the problem is as follows:
lets assume we have the following function body
```
{
Store(...);
... = Sample(...);
}
```
What will happen is, Store sets the inferredCaps to {Fragment, Compute, RayGen, Anyhit, Callable....}.
Then we will get the function Sample, which has the nodeCaps of {Fragment}. 
By logic of Implies, Store does imply Sample, Store has Fragment, it is a super set, this is what `Implies` checks for. 

The problem is that we are not checking for a super set because this means that we will not join the caps to cull for the lowest common denominator. As a result the function's inferred caps will require {Fragment, Compute, RayGen, Anyhit, Callable....} instead of {Fragment}, this is a problem because calling/current functions will allow greater leniency in what can be inside the body. 

For an extended example, we can with the old logic before the change we allow "Fragment only" + "Compute only" functions in a body as long as the first propagated requirement is "Fragment | Compute", this is incorrect behavior.

Reply:
I think it is an error to say (Fragment | Compute) implies (Fragment)?

Reply:
I wonder if the right fix is to fix `implies` instead? That may have other cascading issues, but is it the right thing to do?

Reply:
> I think it is an error to say (Fragment | Compute) implies (Fragment)?
This does correctly return false, that is why we needed to change the logic.

implies works as intended, _I think_.

Reply:
This means the solution is we then do:
join if implies is false & join if implies is true?
if this is correct, then we should just omit implies.